124-exec-engine.Rmd

```diff
@@ -0,0 +1,190 @@
+```
+
+The execution of the chunk will be equivalent to this command line
+
+    <command> <args1> <args> <args2>
+
+Behavior by default, during evaluation will be:
+
+-   Retrieve the path of the command in `command` option.
+-   Write the chunk content into a temporary file `f`
+-   Register the `clean` function to execute on exit, basically to remove the temp file `f`.
+-   Compute the argument of the command line by appending the results of `args1()`, with `args()` then `args2()`.
+
+There should be very rare usage of changing the default behavior, but it can be useful to debug for example or other advanced usage.
+
+You can find some examples below on how to use each option. We are using `Rscript` an d `python` only in the examples below by convenience but this work with any executable file.
+
+# Simple example
+
+`command` can be the name of a command in PATH.
+
+```{exec, command='Rscript'}
+1 + 1
+```
+
+```{exec, command='python'}
+print(1 + 1)
+```
+
+or a full path
+
+```{exec, command = Sys.which('python'), message = FALSE}
+print("Full path command works")
+```
+
+It can also be passed as an option in `engine.opts`.
+
+```{exec, engine.opts = list(command ='python')}
+print(1 + 1)
+```
+
+It is the only option that can be passed as first level chunk option.
+
+# Options
+
+All other options should be provided in the chunk option `engine.opts` in a list. (As seen above, `command` can be optionally in `engine.opts`, too).
+
+They can take character values or functions that returns character values.
+
+## file
+
+It returns the temporary file path that will be used for execution by the command.
+
+It is a function of two argument `function(code, file){}` - By default, function is writing `code` into the temporary file `file` that will be executed by the command. This can be change if needed for other side effect. You would still want to write the `code` into a file (using `xfun::write_utf8()` or another function) otherwise the command will be executed with an empty file.
+
+```{r}
+file2 <- function(code, file) {
+  xfun::write_utf8("print('file2 executed')", file)
+  file
+}
+```
+
+```{exec, engine.opts=list(command = 'python', file = file2)}
+print(1 + 1)
+```
+
+## ext
+
+This will determine the extension to use with the temporary file. By default, the extension will be the name of the command file. Use this option to change it.
+
+For example, this will insure the tempfile has a `.R` extension.
+
+```{exec, engine.opts = list(command='Rscript', ext = 'R')}
+1 + 1
+```
+
+It can be a function of the executable name (i.e `basename(command)`).
+
+```{r, include = FALSE}
+ext_fun <- function(name) {
+  if (grepl("Rscript", name)) return("R")
+  if (grepl("python", name)) return("py")
+  name
+}
+```
+
+```{exec, command = 'python', engine.opts = list(ext = ext_fun)}
+print(1 + 1)
+```
+
+The extension of the temporary does not really matter most of the time but you can use this option if the command you use needs it.
```

Comment:
I am not sure to see what would be the usage of changing `.ext`. 
If you have idea for the example

Reply:
It should be rare to be useful. I have never seen a command that cares about the file extension. I don't think we need an example here.

Reply:
Ok !

Reply:
As I was testing `gcc`, I found that this command does care about the extension. I had to set `ext` to `c`, since the default `.gcc` extension wouldn't work.